### PR TITLE
Update Go compiler version for the go1.x Docker Image template

### PR DESCRIPTION
**Pull request by bmoffatt**
*Description of changes:*

Changes 
* ~~Change the build-image stage to use `public.ecr.aws/amazonlinux/amazonlinux:2`, which maintains a more modern go compiler version in the yum repository~~ 
  * ~~(edit) changed the base build image to `public.ecr.aws/docker/library/golang:1.19`~~ 
  * (edit) updated base build image version to `:1.19`
* ~~Updated the go package dependencies in go.mod, go.sum, by re-running `go mod tidy`~~
  * (edit) already done in #319 
* ~~Changed the final stage of the Docker image to use `public.ecr.aws/amazonlinux/amazonlinux:2` , because the `public.ecr.aws/lambda/go` image does not support `arm64`.~~ 
  * (edit) redo in #322 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: go1.x-image/cookiecutter-aws-sam-hello-golang-lambda-image/{{cookiecutter.project_name}}/hello-world/Dockerfile [hunk 1]
```diff
@@ -1,13 +1,8 @@
-FROM public.ecr.aws/bitnami/golang:1.16 as build-image
-
-WORKDIR /go/src
+FROM public.ecr.aws/amazonlinux/amazonlinux:2 as build-image
+RUN yum install -y golang
```
Comment (lucashuy) [inline go1.x-image/cookiecutter-aws-sam-hello-golang-lambda-image/{{cookiecutter.project_name}}/hello-world/Dockerfile:2]: Would it be possible to use the `1.18` tagged Bitnami golang image over AL2 (ie. `public.ecr.aws/bitnami/golang:1.18`)? The concern here is that the initial time it takes to install golang and it's dependencies with that image isn't ideal from the point of view of a new user that wants to try out golang.

Comment (bmoffatt) [reply go1.x-image/cookiecutter-aws-sam-hello-golang-lambda-image/{{cookiecutter.project_name}}/hello-world/Dockerfile:2]: I think that'll lock out arm image builds 

from: https://gallery.ecr.aws/bitnami/golang
> OS/Arch: Linux, x86-64


What do you think about using https://gallery.ecr.aws/docker/library/golang instead?

Comment (lucashuy) [reply go1.x-image/cookiecutter-aws-sam-hello-golang-lambda-image/{{cookiecutter.project_name}}/hello-world/Dockerfile:2]: That looks like a possible solution, lets try using those images instead

File: go1.x-image/cookiecutter-aws-sam-hello-golang-lambda-image/{{cookiecutter.project_name}}/hello-world/Dockerfile [hunk 2]
```diff
@@ -1,13 +1,7 @@
-FROM public.ecr.aws/bitnami/golang:1.16 as build-image
-
-WORKDIR /go/src
+FROM public.ecr.aws/docker/library/golang:1.19 as build-image
```
Comment (jfuss) [inline go1.x-image/cookiecutter-aws-sam-hello-golang-lambda-image/{{cookiecutter.project_name}}/hello-world/Dockerfile:1]: This uses Go 1.19 but in the `go.mod` file it defines `go 1.18`. Should those match?

File: go1.x-image/cookiecutter-aws-sam-hello-golang-lambda-image/{{cookiecutter.project_name}}/hello-world/Dockerfile [hunk 3]
```diff
@@ -1,13 +1,7 @@
...
-RUN go build -o ../bin
-
-FROM public.ecr.aws/lambda/go:1
-
-COPY --from=build-image /go/bin/ /var/task/
-
-# Command can be overwritten by providing a different command in the template directly.
-CMD ["hello-world"]
+RUN go build -o lambda-handler
+FROM public.ecr.aws/amazonlinux/amazonlinux:2
```
[the remark this answers is not in the document]
Comment (bmoffatt) [reply go1.x-image/cookiecutter-aws-sam-hello-golang-lambda-image/{{cookiecutter.project_name}}/hello-world/Dockerfile:5]: I can change it to use `public.ecr.aws/lambda/provided:al2`

[the remark this answers is not in the document]
Comment (jfuss) [reply go1.x-image/cookiecutter-aws-sam-hello-golang-lambda-image/{{cookiecutter.project_name}}/hello-world/Dockerfile]: Why al2? go1.x is a supported runtime. Moving this to al2 breaks the mapping/expectation we have today. 

If we want to move away from go1.x, we can do that but should be done through SAM CLI not the examples.

[the remark this answers is not in the document]
Comment (sriram-mv) [reply go1.x-image/cookiecutter-aws-sam-hello-golang-lambda-image/{{cookiecutter.project_name}}/hello-world/Dockerfile]: ah! The issue is with the prompts showcased in `sam init`

```
sam init


You can preselect a particular runtime or package type when using the `sam init` experience.
Call `sam init --help` to learn more.

Which template source would you like to use?
	1 - AWS Quick Start Templates
	2 - Custom Template Location
Choice: 1

Choose an AWS Quick Start application template
	1 - Hello World Example
	2 - Multi-step workflow
	3 - Serverless API
	4 - Scheduled task
	5 - Standalone function
	6 - Data processing
	7 - Infrastructure event management
	8 - Serverless Connector Hello World Example
	9 - Multi-step workflow with Connectors
	10 - Lambda EFS example
	11 - Machine Learning
Template: 1

Use the most popular runtime and package type? (Python and zip) [y/N]: N

Which runtime would you like to use?
	1 - aot.dotnet7 (provided.al2)
	2 - dotnet6
	3 - dotnet5.0
	4 - dotnetcore3.1
	5 - go1.x
	6 - go (provided.al2)
	7 - graalvm.java11 (provided.al2)
	8 - graalvm.java17 (provided.al2)
	9 - java11
	10 - java8.al2
	11 - java8
	12 - nodejs18.x
	13 - nodejs16.x
	14 - nodejs14.x
	15 - nodejs12.x
	16 - python3.9
	17 - python3.8
	18 - python3.7
	19 - ruby2.7
	20 - rust (provided.al2)
Runtime: 5

What package type would you like to use?
	1 - Zip
	2 - Image
Package type: 2

```

The provided runtime choice is go1.x, choosing go(al.2) defaults to zip as there as no Image based templates present.

[the remark this answers is not in the document]
Comment (bmoffatt) [reply go1.x-image/cookiecutter-aws-sam-hello-golang-lambda-image/{{cookiecutter.project_name}}/hello-world/Dockerfile]: Oh yeah I've totally goofed that. When I made this PR, I must have misremembered the ordering of the dialog (eg: assumed package type prompt came before the runtime prompt), or hadn't fully digested the structure of the `manifest-v2.json`, or both!!

My main intent @jfuss was to make the Docker image templates for Go usable on ARM, the AL1 -> AL2 bump comes along as an implicit dependency of this. Given my bad assumptions about the `sam init` prompts, I thought that the only option was to change the Dockerfile. Now I think it's clear that I should be duplicating the templates like I did for https://github.com/aws/aws-sam-cli-app-templates/pull/315 - you're right that the `go1.x` runtime, the `public.ecr.aws/lambda/go` images, and AL1 in general, are still supported at least until mid 2023.

[the remark this answers is not in the document]
Comment (bmoffatt) [reply go1.x-image/cookiecutter-aws-sam-hello-golang-lambda-image/{{cookiecutter.project_name}}/hello-world/Dockerfile]: The version bumps for compiler in the build base are still relevant for the `go1.x-image` template. The `FROM public.ecr.aws/bitnami/golang:1.16 as build-image` -> `FROM public.ecr.aws/docker/library/golang:1.19 as build-image`. Would you prefer these changes to land in a separate PR or keep it within the scope of this one?

[the remark this answers is not in the document]
Comment (bmoffatt) [reply go1.x-image/cookiecutter-aws-sam-hello-golang-lambda-image/{{cookiecutter.project_name}}/hello-world/Dockerfile]: After doing more reading, looks like I can't get away with adding the new image template I wanted without also doing a CLI change. It seems [this map](https://github.com/aws/aws-sam-cli/blob/d0f99f2f0160e53f04f08949ac3f779eb6f61f79/samcli/local/common/runtime_template.py#L126-L143) needs to be updated for the [interactive workflow to discover the entry](https://github.com/aws/aws-sam-cli/blob/d0f99f2f0160e53f04f08949ac3f779eb6f61f79/samcli/commands/init/interactive_init_flow.py#L207-L214).

[the remark this answers is not in the document]
Comment (bmoffatt) [reply go1.x-image/cookiecutter-aws-sam-hello-golang-lambda-image/{{cookiecutter.project_name}}/hello-world/Dockerfile]: I'm sure I've still got some aspect of that wrong though :) - so would appreciate you both taking a look at my second attempt: https://github.com/aws/aws-sam-cli-app-templates/pull/322

**Review by sriram-mv — APPROVED** [review]

**Review by mndeveci — APPROVED** [review]
This version of the change LGTM. But looking at all previous discussion, I believe this PR was doing something else rather than just bumping a minor version here.

**Review by jfuss — APPROVED** [review]

**Pull request merged**